### PR TITLE
[Darwin] Properly delete chip::Callback on success/failure

### PR DESCRIPTION
**Pull request by vivien-apple**
#### Problem

When success/failure callbacks are fired, one of them is not properly delete.

#### Change overview
* Glued them together under a single `CHIPCallbackBridge` class

#### Testing
It was tested locally using: `xcodebuild test -target "CHIP" -scheme "CHIP Framework Tests" -sdk macosx`

### CODE REVIEW

File: src/darwin/Framework/CHIP/templates/CHIPClustersObjc-src.zapt [hunk 1]
```diff
@@ -16,131 +16,101 @@
...
-                  @"{{asSymbol label}}": [NSNumber numberWith{{asObjectiveCNumberType label type false}}:{{asSymbol label}}],
-                  {{/if}}
-                  {{/chip_cluster_response_arguments}}
-                });
-                callback->Cancel();
-                delete callback;
-            });
+            chip::Callback::Callback<DefaultFailureCallback> * failureCallback = new chip::Callback::Callback<DefaultFailureCallback>(OnFailureFn, this);
+            Failure = failureCallback->Cancel();
         }
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/templates/CHIPClustersObjc-src.zapt:29]: Maybe add a comment about how subclasses should initialize `Success`?

File: src/darwin/Framework/CHIP/templates/CHIPClustersObjc-src.zapt [hunk 2]
```diff
@@ -16,131 +16,101 @@
...
-    ResponseHandler mHandler;
-    dispatch_queue_t mQueue;
-};
 
-{{/chip_cluster_responses}}
-{{/chip_client_clusters}}
+    ~CHIPCallbackBridge()
+    {
+        Success->Cancel();
+        Failure->Cancel();
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/templates/CHIPClustersObjc-src.zapt:34]: Do you need the explicit `Cancel` calls?  I thought the whole point of using `Cancelable` is that its destructor cancels?

File: src/darwin/Framework/CHIP/templates/CHIPClustersObjc-src.zapt [hunk 3]
```diff
@@ -16,131 +16,101 @@
...
+        delete Success;
+        delete Failure;
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/templates/CHIPClustersObjc-src.zapt:36]: Just use `unique_ptr` for the members?

File: src/darwin/Framework/CHIP/templates/CHIPClustersObjc-src.zapt [hunk 4]
```diff
@@ -174,10 +144,40 @@ private:
...
+{
+  if ([value isKindOfClass:[NSData class]])
+  {
+      NSData * v = (NSData *) value;
+      return chip::ByteSpan((const uint8_t*)v.bytes, v.length);
+  }
+  else
+  {
+      NSString * v = (NSString *) value;
+      return chip::ByteSpan((const uint8_t*)[v dataUsingEncoding:NSUTF8StringEncoding].bytes, [v lengthOfBytesUsingEncoding:NSUTF8StringEncoding]);
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/templates/CHIPClustersObjc-src.zapt:178]: Does the return value from `dataUsingEncoding` stay alive after this function returns?  We're returning a ByteSpan pointing into it.....

I would not have expected it to survive, myself.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: src/darwin/Framework/CHIP/templates/CHIPClustersObjc-src.zapt (hunk 1 above)
Comment (vivien-apple) [reply src/darwin/Framework/CHIP/templates/CHIPClustersObjc-src.zapt]: I have updated the code a lot. I don't think the comment is necessary now.

File: src/darwin/Framework/CHIP/templates/CHIPClustersObjc-src.zapt (hunk 3 above)
Comment (vivien-apple) [reply src/darwin/Framework/CHIP/templates/CHIPClustersObjc-src.zapt]: I have removed the `new`, `delete` call for the success/failure callbacks.

File: src/darwin/Framework/CHIP/templates/CHIPClustersObjc-src.zapt (hunk 4 above)
Comment (vivien-apple) [reply src/darwin/Framework/CHIP/templates/CHIPClustersObjc-src.zapt]: My understanding is yes. The value returned by `dataUsingEncoding` is an `autoreleased` value that will be drained once the event loop has pass is over.
The current code does use the `ByteSpan` into a `dispatch_sync` call to the stack, and I would expect the underlying data to still be alive.

**Review by bzbarsky-apple — APPROVED** [review]

**Review by andy31415 — APPROVED** [review]
rubberstamp

**Review by woody-apple — APPROVED** [review]

**Review by jelderton — APPROVED** [review]

**Comment by andy31415** [comment]
@vivien-apple - conflicts. adding this comment to also remove stale flag (if  possible)

**Pull request merged**